## Problem Set 2: Merging and regular expressions

**Total points (without extra credit)**: 30 

**Background on the policy context**: here, we're going to use two datasets to practice reshaping, merging, and regular expression patterns. Both datasets relate to the broader issue of which employers might be violating the rights of temporary guestworkers granted visas under the H-2A program. Here are some articles about potential exploitation of guestworkers by firms and inequality caused by minimal oversight:

- News media coverage of labor abuses of temporary guestworkers: https://www.buzzfeednews.com/article/kenbensinger/the-pushovers 
- GAO report on labor abuses of temporary guestworkers: https://www.gao.gov/products/gao-15-154

The following datasets are located in `pset2_inputdata` (need to unzip): 

- `jobs_clean`: a dataset of guestworker jobs posted by many employers, some of whom have been debarred (banned) from the program for labor abuses; others not debarred
- `debar`: a dataset of employers who committed violations of labor regulations meant to protect temporary guestworkers 


You can view a codebook here: https://docs.google.com/spreadsheets/d/1rF9GJEC8pPKxipD0TsoG9DVdqz3EJ-b-BHEtyioAX7I/edit?usp=sharing


In [9]:
## helpful packages
import pandas as pd
import numpy as np
import random
import re
import os

## repeated printouts
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


# 1. Reshaping data (13 points total)

Load the following dataset stored in `pset2_inputdata`: `debar.csv`

This represents employers temporarily banned from hiring workers (debar.csv); call this `debar`


View the head()


In [10]:
##load in the dataset and call it debar
debar  = pd.read_csv('debar.csv')

#view the head
debar.head(10)



Name      City, State  \
0                         J&J Harvesting        Leads, ND   
1                 Stahlman Apiaries, Inc        Selby, SD   
2                          Trust Nursery      Pulaski, NY   
3                  Anton Fertilizer Inc.      Dighton, KS   
4       Great Plains Fluid Service, Inc.   Greensburg, KS   
5                            Promax Inc.   Whitewater, KS   
6                Annabella Land & Cattle    Annabella, UT   
7                    Autumn Hill Orchard       Groton, MA   
8  Caddo Creek Ranch, dba Paradise Ranch        Caddo, TX   
9                         Yolanda Chavez  Santa Maria, CA   

                                        Violation Duration Start date  \
0  Failure to respond to audit (partial response)  2 years  1/19/2014   
1  Failure to respond to audit (partial response)   1 year  2/19/2015   
2  Failure to respond to audit (partial response)   1 year  3/21/2014   
3       Failure to respond to audit (no response)  2 years  3/30/2014   
4       Failure to respond to audit (no response)  2 years  3/30/2014   
5                    Failure to Hire U.S. workers  2 years  5/15/2014   
6                                     Non Payment   1 year   5/9/2014   
7       Failure to respond to audit (no response)  2 years   7/6/2014   
8  Failure to respond to audit (partial response)  2 years  7/20/2014   
9                                     Non Payment   1 year  7/23/2014   

    End date  
0  1/18/2016  
1  2/14/2016  
2  3/20/2015  
3  3/29/2016  
4  3/29/2016  
5  5/14/2016  
6   5/9/2015  
7   7/5/2016  
8  7/19/2016  
9  7/22/2015

## 1.1 (1 point)

Print the number of rows in `debar` versus the number of unique employer names (`Name`). Is there one row per employer or multiple rows for some employers?

In [11]:
## use len in order to print the number of rows in debar
len(debar)

#use .nunique() to print the number of unique employer names
debar.Name.nunique()


114

98

There are 114 rows and only 98 unique employer names, therefore there are multiple rows for some employers.

## 1.2 Investigating duplicated rows (2 points)

A. Create a new column in `debar`--`is_repeated`-- that tells us whether an employer (`Name`) is repeated > 1 times

*Hint*: there are multiple ways to solve this but some possibilities to get the list of names that are repeated are:
- Using value_counts() on the `Name` variable and extracting the index from that value counts 
- Using groupby to count the rows attached to one name

B. Print the rows where `is_repeated == True` and interpret

C. Subset to the rows where `is_repeated == True` and save that data as `mult_debar`. Print the head() and shape

In [12]:
#A 

#counts how many times each value appears in the name column 
name_counts = debar['Name'].value_counts().reset_index()

#renames the columns in name_counts
name_counts.columns = ['Name', 'count']

#update the df
debar = debar.merge(name_counts, on='Name', how='left')

#add new column in debar
debar['is_repeated'] = debar['count'] > 1

debar.head(10)


Name      City, State  \
0                         J&J Harvesting        Leads, ND   
1                 Stahlman Apiaries, Inc        Selby, SD   
2                          Trust Nursery      Pulaski, NY   
3                  Anton Fertilizer Inc.      Dighton, KS   
4       Great Plains Fluid Service, Inc.   Greensburg, KS   
5                            Promax Inc.   Whitewater, KS   
6                Annabella Land & Cattle    Annabella, UT   
7                    Autumn Hill Orchard       Groton, MA   
8  Caddo Creek Ranch, dba Paradise Ranch        Caddo, TX   
9                         Yolanda Chavez  Santa Maria, CA   

                                        Violation Duration Start date  \
0  Failure to respond to audit (partial response)  2 years  1/19/2014   
1  Failure to respond to audit (partial response)   1 year  2/19/2015   
2  Failure to respond to audit (partial response)   1 year  3/21/2014   
3       Failure to respond to audit (no response)  2 years  3/30/2014   
4       Failure to respond to audit (no response)  2 years  3/30/2014   
5                    Failure to Hire U.S. workers  2 years  5/15/2014   
6                                     Non Payment   1 year   5/9/2014   
7       Failure to respond to audit (no response)  2 years   7/6/2014   
8  Failure to respond to audit (partial response)  2 years  7/20/2014   
9                                     Non Payment   1 year  7/23/2014   

    End date  count  is_repeated  
0  1/18/2016      1        False  
1  2/14/2016      1        False  
2  3/20/2015      1        False  
3  3/29/2016      1        False  
4  3/29/2016      1        False  
5  5/14/2016      1        False  
6   5/9/2015      2         True  
7   7/5/2016      2         True  
8  7/19/2016      2         True  
9  7/22/2015      1        False

In [13]:
#B
#prints just the rows where is_repeated is true
debar[debar["is_repeated"] == True]


Name       City, State  \
6                    Annabella Land & Cattle     Annabella, UT   
7                        Autumn Hill Orchard        Groton, MA   
8      Caddo Creek Ranch, dba Paradise Ranch         Caddo, TX   
11                     Loewen Harvesting LLC   Brownsville, TX   
12               Rollo Farm Labor Contractor         Miami, FL   
14                             Sharon Mathis        Tifton, GA   
15                                 SRT Farms        Morton, TX   
16                               Mark Duncan     Roosevelt, UT   
17          Maple Ridge Custom Services, LLC     Altheimer, AK   
18                                 F&W Farms       Ingalls, KS   
19                        Cisco Produce Inc.         Cairo, GA   
21   Old Tree Farms/Verpaalen Custom Service         Volga, SD   
24               Rollo Farm Labor Contractor         Miami, FL   
25                     Loewen Harvesting LLC    Brownfield, TX   
28     Caddo Creek Ranch, dba Paradise Ranch      Caddo, Texas   
29                       Autumn Hill Orchard        Groton, MA   
30                   Annabella Land & Cattle   Annabella, Utah   
31                      Macky and Brad Farms        Plains, TX   
51   Old Tree Farms/Verpaalen Custom Service         Volga, SD   
55                      Macky and Brad Farms        Plains, TX   
56                        Cisco Produce Inc.         Cairo, GA   
58                                 F&W Farms       Ingalls, KS   
59          Maple Ridge Custom Services, LLC     Altheimer, AR   
60                               Mark Duncan     Roosevelt, UT   
61                                 SRT Farms        Morton, TX   
62                             Sharon Mathis        Tifton, GA   
73                              Xavier Horne    Lyons, Georgia   
89                              Xavier Horne    Lyons, Georgia   
103                         Dove Creek Farms  Mount Vernon, TX   
106                             Turner Farms         Healy, KS   
109                         Dove Creek Farms  Mount Vernon, TX   
111                             Turner Farms         Healy, KS   

                                             Violation  Duration  Start date  \
6                                          Non Payment    1 year    5/9/2014   
7            Failure to respond to audit (no response)   2 years    7/6/2014   
8       Failure to respond to audit (partial response)   2 years   7/20/2014   
11      Failure to respond to audit (partial response)    1 year   8/20/2014   
12           Failure to respond to audit (no response)   2 years   8/23/2014   
14           Failure to respond to audit (no response)   2 years  11/16/2014   
15           Failure to respond to audit (no response)   2 years  11/16/2014   
16           Failure to respond to audit (no response)   2 years  11/16/2014   
17      Failure to respond to audit (partial response)   2 years  11/16/2014   
18      Failure to respond to audit (partial response)   2 years  12/10/2014   
19           Failure to respond to audit (no response)   2 years  12/10/2014   
21                                       WHD Debarment   3 years  12/11/2014   
24          Impeding the Audit Process – Non- Response   2 years   8/23/2014   
25      Impeding the Audit Process – Partial- Response    1 year   8/20/2014   
28      Impeding the Audit Process – Partial- Response   2 years   7/20/2014   
29          Impeding the Audit Process – Non- Response   2 years    7/6/2014   
30                                         Non-payment    1 year    5/9/2014   
31           Failure to respond to audit (no response)    1 year   2/13/2015   
51                                 Wage Hour Debarment   3 years   12/1/2014   
55      Impeding the Audit Process – Partial- Response    1 year   2/13/2015   
56          Impeding the Audit Process – Non- Response   2 years  12/10/2015   
58      Impeding the Audit Process – Partial- Response    1 year  12/10/2014   
59      Impeding the 

INTERPRETATION:

-many (but not all) of the repeated entries are from multiple violations 

-the ones that are not from mulitple violations seem to just be repeats of the same entry


In [14]:
#C

#subsets the rows and saves it to mult_debar
mult_debar = debar[debar['is_repeated'] == True]

#print the head
mult_debar.head()

#print the shape
mult_debar.shape


Name      City, State  \
6                 Annabella Land & Cattle    Annabella, UT   
7                     Autumn Hill Orchard       Groton, MA   
8   Caddo Creek Ranch, dba Paradise Ranch        Caddo, TX   
11                  Loewen Harvesting LLC  Brownsville, TX   
12            Rollo Farm Labor Contractor        Miami, FL   

                                         Violation Duration Start date  \
6                                      Non Payment   1 year   5/9/2014   
7        Failure to respond to audit (no response)  2 years   7/6/2014   
8   Failure to respond to audit (partial response)  2 years  7/20/2014   
11  Failure to respond to audit (partial response)   1 year  8/20/2014   
12       Failure to respond to audit (no response)  2 years  8/23/2014   

     End date  count  is_repeated  
6    5/9/2015      2         True  
7    7/5/2016      2         True  
8   7/19/2016      2         True  
11  8/19/2015      2         True  
12  8/22/2016      2         True

(32, 8)

## 1.3 Reshape mult_debar to wide to begin filtering out duplicates (4 points)

You want to separate out two cases:

- Cases where the repeat rows for one employer are due to duplicated data 
- Cases where the repeat rows for one employer represent repeated violations for different issues

There are various ways to check duplicates in this data (eg converting `Violation` to lowercase; replacing spelled-out states with two-dig state codes)

We're going to use the simple rule of:

- A row is a duplicate if, within an employer (defined by Name + City, State), the Start date for each row's violation is the same 

To begin to check this, reshape `mult_debar` to a wide dataframe (`mult_debar_wide`) with the following columns, treating the `Name` and `City, State` as the index for the pivot:

- Name
- City, State
- start_date_viol1
- start_date_viol2

Print the head and shape

In [15]:
#since duplicate = if 'Employer' (Name + City,State) and 'Start date' are equal
mult_debar["each_violation"] = mult_debar.groupby(['Name', 'City, State'])

#reshape mult_debar to a wide df 
mult_debar_wide = mult_debar.pivot_table(index=['Name', 'City, State'],
    columns='each_violation',
    values='Start date',
    aggfunc='first')

#prints the head and shape
mult_debar_wide.head(10)
mult_debar_wide.shape



#dictionary


ValueError: Length of values (20) does not match length of index (32)

In [ ]:

# Dictionary to store information
employer_dict = {}

# List to collect duplicate info
debar['duplicate_key'] = None  # Initialize a new column for duplicate marking

for index, row in debar.iterrows():
    key = (row['Name'], row['City, State'])
    start_date = row['Start date']
    
    if key not in employer_dict:
        employer_dict[key] = {}
    
    if start_date in employer_dict[key]:
        # If start date is already recorded, it's a duplicate
        debar.loc[index, 'duplicate_key'] = 'duplicate'
        employer_dict[key][start_date] += 1
    else:
        # Record the start date for this employer
        employer_dict[key][start_date] = 1
        debar.loc[index, 'duplicate_key'] = 'unique'





## 1.4 Filter out duplicates from original debar data (6 points)

A. Using `mult_debar_wide`, add a column `is_dup` that takes value of True for cases where start_date_viol1 == start_date_viol2 marking the row as a duplicate

B. Going back to the original long-format data you loaded at the beginning- `debar`
    - For employers where `is_dup == True` as indicated by your wide-format dataframe, only keep `violnum == viol1`
    - For all other employers (so is_dup == False and ones we didnt need to check duplicates for), keep all violnum
    - Remove the `is_repeated` column from the `debar` data

**Hint**: you can complete part B without a for loop; `pd.concat` with axis = 0 (row binding) is one way

Call the resulting dataframe `debar_clean` and print the shape and # of unique employer names

In [ ]:
## your code here

# 2. Merging and regex (17 points total)



## 2.1 Load data on job postings

The previous dataset contains a small subset of employers who faced temporary bans due to violations of H-2A program regulations

Since most of the bans have expired, we're going to see which of those employers posted new H-2A jobs in the first quarter of 2021 

Loading the `jobs_clean.csv` data stored in `pset4_inputdata`

In [19]:
##load in the second dataset and call it job_clean
jobs  = pd.read_csv('jobs.csv')

#view the head
jobs.head(10)



CASE_NUMBER                                     CASE_STATUS  \
0  H-300-20199-721302                Determination Issued - Withdrawn   
1  H-300-20231-773906            Determination Issued - Certification   
2  H-300-20231-774123            Determination Issued - Certification   
3  H-300-20231-774151            Determination Issued - Certification   
4  H-300-20231-774508            Determination Issued - Certification   
5  H-300-20231-775063            Determination Issued - Certification   
6  H-300-20232-777530            Determination Issued - Certification   
7  H-300-20233-780540  Determination Issued - Certification (Expired)   
8  H-300-20234-782276            Determination Issued - Certification   
9  H-300-20235-782945            Determination Issued - Certification   

             RECEIVED_DATE            DECISION_DATE  \
0  2020-07-17 14:50:40.840  2020-10-01 00:00:00.000   
1  2020-08-20 10:38:15.620  2020-10-01 00:00:00.000   
2  2020-08-24 15:33:14.340  2020-10-01 00:00:00.000   
3  2020-08-21 12:08:09.760  2020-10-01 00:00:00.000   
4  2020-08-20 10:17:34.530  2020-10-01 00:00:00.000   
5  2020-08-21 11:36:32.360  2020-10-01 00:00:00.000   
6  2020-08-31 14:14:31.920  2020-10-01 00:00:00.000   
7  2020-09-04 14:26:00.270  2020-10-01 00:00:00.000   
8  2020-08-31 13:12:17.840  2020-10-01 00:00:00.000   
9  2020-09-17 11:04:44.130  2020-10-01 00:00:00.000   

   TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR NATURE_OF_TEMPORARY_NEED  \
0           Individual Employer                    N                 Seasonal   
1           Association - Agent                    N                 Seasonal   
2           Individual Employer                    N                 Seasonal   
3           Individual Employer                    N                 Seasonal   
4           Individual Employer                    Y                 Seasonal   
5           Individual Employer                    N                 Seasonal   
6           Individual Employer                    N                 Seasonal   
7  Association - Joint Employer                    N                 Seasonal   
8           Individual Employer                    N                 Seasonal   
9           Individual Employer                    N                 Seasonal   

  EMERGENCY_FILING                             EMPLOYER_NAME  \
0                Y        Fazio Farms Operating Company, LLC   
1                N                        Charlie Sunderland   
2                N                         Michael Rudebusch   
3                N                              Lodahl Farms   
4                N                   Dunson Harvesting, Inc.   
5                N           CUMBERLAND VALLEY NURSERIES INC   
6                N                 O'Connor Ranch Lands, LLC   
7                N  Farm Labor Association for Growers, Inc.   
8                N                                Rksk Farms   
9                N                             CY FARMS, LLC   

                TRADE_NAME_DBA  ... ADDENDUM_B_HOUSING_ATTACHED  \
0                          NaN  ...                           N   
1  Panter & Sunderland Nursery  ...                           N   
2                          NaN  ...                           N   
3                          NaN  ...                           Y   
4      Dunson Harvesting, Inc.  ...                           Y   
5                          NaN  ...                           N   
6                          NaN  ...                           N   
7                          NaN  ...                           N   
8                          NaN  ...                           N   
9                          NaN  ...                           N   

  TOTAL_HOUSING_RECORDS MEALS_PROVIDED MEALS_CHARGED  \
0                     1              Y         12.68   
1                     1              N           NaN   
2                     1              N           NaN   
3                     2              N   

##  2.2 Try inner join on employer name  (2 points)

- Use the `EMPLOYER_NAME` field of the `jobs` dataset
- Use the `Name` field of the `debar_clean` dataset 

A. Use pd.merge with an inner join on those fields to see whether there are any exact matches. 

B. If there are exact matches, print the row(s) with exact matches



In [ ]:
## your code here

## 2.3 Targeted regex (10 points total)

You want to see if you can increase the exact match rate with some basic cleaning of each 
of the employer name fields in each dataset 

### 2.3.1 Converting to upper (2 points)

A. Convert the `EMPLOYER_NAME` and `Name` fields to uppercase using list comprehension rather than df.varname.str.upper() (it's fine to do a separate list comprehension line for each of the two columns)

B. Print a random sample of 15 values of each result

C. Assign the full vector of uppercase names back to the original data, writing over the original `EMPLOYER_NAME` and `Name` columns 


In [ ]:
## insert your code to turn into uppercase here


In [ ]:
## insert your code for the random sample

In [ ]:
## insert your code for assigning the uppercase names back to the data

### 2.3.2 Cleaning up punctuation (4 points)

You notice that INC, CO, and LLC are sometimes followed by a period (.) but sometimes not

A. For each dataset, write a regex pattern using `re.sub` to remove the . but only if it's preceded by INC, LLC, or CO 

Make sure LLC, INC, CO remain part of the string but just without the dot

B. Test the pattern on the positive and negative example we provide below and print the result. See the Github issue for examples of what to return


**Hint**: https://stackoverflow.com/questions/7191209/python-re-sub-replace-with-matched-content



In [ ]:
pos_example_1 = "CISCO PRODUCE INC."
pos_example_2 = "AVOYELLES HONEY CO., LLC"
neg_example = "E.V. RANCH LLP"

In [ ]:
## insert your code here with the regex pattern for part A

## insert your code to use re.sub to apply the pattern to the test cases for part B

### 2.3.3 (4 points)

Use that pattern in conjunction with `re.sub` and list comprehension to clean the employer name columns in each dataset. Save the new columns as `name_clean` in each. Then, use row subsetting to (1) subset to rows that changed names and (2) for:

- `debar_clean` print the `Name` and `name_clean` columns
- `jobs` print the `EMPLOYER_NAME` and `name_clean` columns

Make sure to use the uppercase versions of the variables


In [ ]:
## your code here to clean the columns

In [ ]:
## your code here to print the head

## 2.4 More joins and more cleaning (5 points)

A. Conduct another inner join between `jobs` and `debar_clean` now using the `name_clean` column; print the result. Did the cleaning result in any more employers matched between the two datasets?

B. Create a new column in `debar_clean` called `name_clean_2` that uses regex to take the following name in that dataset:

- `SLASH E.V. RANCH LLP` in the `debar_clean` dataset

And cleans it up so that it matches with this employer in `jobs`

- `SLASH EV RANCH` in the `jobs` dataset

Eg a pattern to remove the dots in the EV and the space+LLP-- you can apply the pattern to all employer names in debar_clean (so don't need to worry about only applying it to that one employer)


C. Conduct a left join using `name_clean_2` as the join column where the left hand dataframe is `jobs`; right hand dataframe is `debar_clean`, store the result as a dataframe, and print the rows where the merge indicator indicates the row was found in both dataframe

**Note**: this manual cleaning process is inefficient and helps motivate why talked about fuzzy matching. Fuzzy matching could recognize that Slash EV ranch is a highly similar string to slash ev ranch llp and match them without us needing to use regex to make the strings identical.

In [ ]:
## your code here

# 3. Optional extra credit 1: regex to separate companies from individuals (1 point)

You notice some employers in `debar_clean` have both the name of the company and the name of individual, e.g.:
    
COUNTY FAIR FARM (COMPANY) AND ANDREW WILLIAMSON (INDIVIDUAL)*

Use the uppercase/cleaned `name_clean` in `debar_clean`

A. Write a regex pattern that does the following:
    - Captures the pattern that occurs before COMPANY if (COMPANY) is in string; so in example above, extracts COUNTY FAIR FARM 
    - Captures the pattern that occurs before INDIVIDUAL if (INDIVIDUAL) is also in string -- so in above, extracts ANDREW WILLIAMSON (so omit the "and")
    
B. Test the pattern on `pos_example` and `neg_example`-- make sure former returns a list (if using find.all) or match object (if using re.search) with the company name and individual name separated out; make sure latter returns empty
    
**Hints and resources**: for step A, you can either use re.search, re.match, or re.findall; don't worry about matching B&R Harvesting and Paul Cruz (Individual)

- Same regex resources as above
    

In [68]:
import re

pos_example = "COUNTY FAIR FARM (COMPANY) AND ANDREW WILLIAMSON (INDIVIDUAL)*"
neg_example = "CISCO PRODUCE INC"

## your code here to define the pattern
pattern = r"(.+?) \s*\((COMPANY)\)\s*AND\s*(.+?)\s*\((INDIVIDUAL)\)"

## your code here to apply it to the positive example
pos = re.search(pattern, pos_example)
#print(pos) checks that it returns a match object
print("Company: ", pos.group(1), "Individual: ", pos.group(3))

## your code here to apply it to the negative example
neg = re.search(pattern, neg_example)
print(neg)


Company:  COUNTY FAIR FARM Individual:  ANDREW WILLIAMSON
None


C. Iterate over the `name_clean` column in debar and use regex to create two new columns in `debar_clean`:
   - `co_name`: A column for company (full `name_clean` string if no match; pattern before COMPANY if one extracted)
   - `ind_name`: A column for individual (full `name_clean` string if no match; pattern before INDIVIDUAL if one extracted)
 


In [69]:
# Defines a function that applies regex and extracts company and individual names
def extract_names(text):
    match = re.search(pattern, text) #finds matches for regex pattern
    if match:
        #if match is found extracts first the company name and then the individual
        company = match.group(1)
        individual = match.group(2)
    #returns a pandas series containing company and individual name   
    return pd.Series([company, individual]) 
    
# Apply the function to each row in the name_clean column
debar_clean[['co_name', 'ind_name']] = debar_clean['name_clean'].apply(extract_names)
debar_clean[['co_name', 'ind_name']].head()

NameError: name 'debar_clean' is not defined

   
D. Print three columns for the rows in `debar_clean` containing the negative example and positive example described above (county fair farm and cisco produce):

- `name_clean`
- `co_name`
- `ind_name`
- `Violation`

**Note**: as shown in the outcome there may be duplicates of the same company reflecting different violations

In [ ]:
# your code here
names_to_check = ['COUNTY FAIR FARM (COMPANY) AND ANDREW WILLIAMSON (INDIVIDUAL)*' "CISCO PRODUCE INC" ]

debar_clean[debar_clean ['name_clean' ].isin(names_to_check)] [['name_clean', 'co_name', 'ind_name', 'Violation']]

# 4. Optional extra credit 2 (up to 3 points)

- For 1 point extra credit, create a visualization with 1+ of the existing fields in either the raw `jobs` or `debar` data. We'll be showing cool visualizations in class so use your imagination! Options could include visualizing between-state or over-time variation

- For 3 points extra credit instead, geocode the employer addresses in `jobs` and plot the addresses of jobs as points overlaid on top of a map of Georgia 
    - **Note**: this extra credit involves Googling since we have not yet covered spatial data. 
        - For discussion of how to geocode addresses -> lat/long, see: https://www.natasshaselvaraj.com/a-step-by-step-guide-on-geocoding-in-python/ 
        - For discussion of plotting lat/long dots against a map, see this discussion of geopandas: https://towardsdatascience.com/plotting-maps-with-geopandas-428c97295a73
    - Relevant columns include `EMPLOYER_ADDRESS_1` 
    - The geocoding might have a long runtime so feel free to implement it in a separate .py script that you submit alongside your notebook and to just read in the geocoded data

Plan: 

Preprocess the Data:
Identify and extract the address-related columns.
Clean and prepare these addresses for geocoding.

                                 Setup Google Maps API for Geocoding:
Ensure you have an API key for Google Maps.
Initialize the Google Maps client.

                          Geocode the Addresses:
Apply the geocoding process to each address.
Handle any errors or non-geocodable addresses gracefully.

                          Store and Review the Results:
Save the latitude and longitude back into the DataFrame.
Check the output for correctness and completeness.

In [20]:
##geocoding the employer addresses in jobs

#installing 
import csv
import pandas as pd
import googlemaps


In [38]:
#check the head of the df
jobs.head()

#find the columns with relevant data
jobs.columns
address_keywords = ['address', 'street', 'city', 'state', 'zip', 'postal', 'location', 'country', 'region', 'district']

# Filter columns that contain any of the keywords
address_columns = [col for col in jobs.columns if any(keyword in col.lower() for keyword in address_keywords)]
print(address_columns)


CASE_NUMBER                           CASE_STATUS  \
0  H-300-20199-721302      Determination Issued - Withdrawn   
1  H-300-20231-773906  Determination Issued - Certification   
2  H-300-20231-774123  Determination Issued - Certification   
3  H-300-20231-774151  Determination Issued - Certification   
4  H-300-20231-774508  Determination Issued - Certification   

             RECEIVED_DATE            DECISION_DATE  \
0  2020-07-17 14:50:40.840  2020-10-01 00:00:00.000   
1  2020-08-20 10:38:15.620  2020-10-01 00:00:00.000   
2  2020-08-24 15:33:14.340  2020-10-01 00:00:00.000   
3  2020-08-21 12:08:09.760  2020-10-01 00:00:00.000   
4  2020-08-20 10:17:34.530  2020-10-01 00:00:00.000   

  TYPE_OF_EMPLOYER_APPLICATION H2A_LABOR_CONTRACTOR NATURE_OF_TEMPORARY_NEED  \
0          Individual Employer                    N                 Seasonal   
1          Association - Agent                    N                 Seasonal   
2          Individual Employer                    N                 Seasonal   
3          Individual Employer                    N                 Seasonal   
4          Individual Employer                    Y                 Seasonal   

  EMERGENCY_FILING                       EMPLOYER_NAME  \
0                Y  Fazio Farms Operating Company, LLC   
1                N                  Charlie Sunderland   
2                N                   Michael Rudebusch   
3                N                        Lodahl Farms   
4                N             Dunson Harvesting, Inc.   

                TRADE_NAME_DBA  ... ADDENDUM_B_HOUSING_ATTACHED  \
0                          NaN  ...                           N   
1  Panter & Sunderland Nursery  ...                           N   
2                          NaN  ...                           N   
3                          NaN  ...                           Y   
4      Dunson Harvesting, Inc.  ...                           Y   

  TOTAL_HOUSING_RECORDS MEALS_PROVIDED MEALS_CHARGED  \
0                     1              Y         12.68   
1                     1              N           NaN   
2                     1              N           NaN   
3                     2              N           NaN   
4                     8              N           NaN   

  MEAL_REIMBURSEMENT_MINIMUM MEAL_REIMBURSEMENT_MAXIMUM PHONE_TO_APPLY  \
0                      12.68                       55.0    13607017661   
1                      12.68                       55.0    19318083783   
2                      12.68                       55.0    19369333827   
3                      12.68                       55.0    14069637560   
4                      12.68                       55.0    18632939888   

                  EMAIL_TO_APPLY                             WEBSITE_TO_APPLY  \
0           faziofarms@gmail.com                                          NaN   
1                            NaN  https://www.jobs4tn.gov/vosnet/Default.aspx   
2  fayethlynpitre@rocketmail.com                                          NaN   
3        lodahl_kelsey@yahoo.com                                          NaN   
4                            NaN                        www.employflorida.com   

   TOTAL_ADDENDUM_A_RECORDS  
0                         0  
1                         0  
2                         0  
3                         0  
4                         4  

[5 rows x 138 columns]

Index(['CASE_NUMBER', 'CASE_STATUS', 'RECEIVED_DATE', 'DECISION_DATE',
       'TYPE_OF_EMPLOYER_APPLICATION', 'H2A_LABOR_CONTRACTOR',
       'NATURE_OF_TEMPORARY_NEED', 'EMERGENCY_FILING', 'EMPLOYER_NAME',
       'TRADE_NAME_DBA',
       ...
       'ADDENDUM_B_HOUSING_ATTACHED', 'TOTAL_HOUSING_RECORDS',
       'MEALS_PROVIDED', 'MEALS_CHARGED', 'MEAL_REIMBURSEMENT_MINIMUM',
       'MEAL_REIMBURSEMENT_MAXIMUM', 'PHONE_TO_APPLY', 'EMAIL_TO_APPLY',
       'WEBSITE_TO_APPLY', 'TOTAL_ADDENDUM_A_RECORDS'],
      dtype='object', length=138)

['EMPLOYER_ADDRESS_1', 'EMPLOYER_ADDRESS_2', 'EMPLOYER_CITY', 'EMPLOYER_STATE', 'EMPLOYER_POSTAL_CODE', 'EMPLOYER_COUNTRY', 'EMPLOYER_POC_ADDRESS1', 'EMPLOYER_POC_ADDRESS2', 'EMPLOYER_POC_CITY', 'EMPLOYER_POC_STATE', 'EMPLOYER_POC_POSTAL_CODE', 'EMPLOYER_POC_COUNTRY', 'ATTORNEY_AGENT_ADDRESS_1', 'ATTORNEY_AGENT_ADDRESS_2', 'ATTORNEY_AGENT_CITY', 'ATTORNEY_AGENT_STATE', 'ATTORNEY_AGENT_POSTAL_CODE', 'ATTORNEY_AGENT_COUNTRY', 'STATE_OF_HIGHEST_COURT', 'NAME_OF_HIGHEST_STATE_COURT', 'WORKSITE_ADDRESS', 'WORKSITE_CITY', 'WORKSITE_STATE', 'WORKSITE_POSTAL_CODE', 'HOUSING_ADDRESS_LOCATION', 'HOUSING_CITY', 'HOUSING_STATE', 'HOUSING_POSTAL_CODE', 'HOUSING_COMPLIANCE_STATE']


In [40]:
#extract the relevant columns (address) to create a new df
address_columns = ['EMPLOYER_ADDRESS_1', 'EMPLOYER_ADDRESS_2', 'EMPLOYER_CITY', 'EMPLOYER_STATE',
    'EMPLOYER_POSTAL_CODE', 'EMPLOYER_COUNTRY']

#create new df
georgia_addresses = jobs[jobs['EMPLOYER_STATE'] == 'GA'][address_columns]

georgia_addresses


EMPLOYER_ADDRESS_1 EMPLOYER_ADDRESS_2  EMPLOYER_CITY  \
14    534 McLain Cemetery Road                NaN          Lyons   
39        9465 Miami Valley Rd                NaN    Fort Valley   
75            30 Sand Creek Rd                NaN         Tifton   
100           30 Sand Creek Rd                NaN         Tifton   
102               19 Lupe Lane                NaN           Wray   
...                        ...                ...            ...   
2455      199 CRAWLEY GAP ROAD                NaN    BLAIRSVILLE   
2563               5753 Hwy 91                NaN  Donalsonville   
2598       1265 US Hwy 82 West                NaN       Leesburg   
2650       1487 Black Dirt Rd                 NaN    Whitesburg    
2692    241 Frank Hutcheson Rd                NaN         Adrian   

     EMPLOYER_STATE EMPLOYER_POSTAL_CODE          EMPLOYER_COUNTRY  
14               GA                30436  UNITED STATES OF AMERICA  
39               GA                31030  UNITED STATES OF AMERICA  
75               GA                31793  UNITED STATES OF AMERICA  
100              GA                31793  UNITED STATES OF AMERICA  
102              GA                31798  UNITED STATES OF AMERICA  
...             ...                  ...                       ...  
2455             GA                30514  UNITED STATES OF AMERICA  
2563             GA                39845  UNITED STATES OF AMERICA  
2598             GA                31763  UNITED STATES OF AMERICA  
2650             GA                30185  UNITED STATES OF AMERICA  
2692             GA                31002  UNITED STATES OF AMERICA  

[74 rows x 6 columns]

In [41]:
#my API key: AIzaSyCK3G3AXvZ-x8HcWrusKZyfz4IMyttwl80

gmaps_key = googlemaps.Client(key="AIzaSyCK3G3AXvZ-x8HcWrusKZyfz4IMyttwl80")

In [45]:
geocode_result = gmaps_key.geocode('1')
print(geocode_result)


ApiError: REQUEST_DENIED (This API project is not authorized to use this API.)

In [ ]:
##plotting the addresses of jobs as points on a map of Georgia

import pandas as pd
import numpy as np
import geopandas as gpd
import matplotlib.pyplot as plt
from shapely.geometry import Point, Polygon